### DISCUSSION TITLE
Feedback: Consider making an IDE for lyte2D similar to amulet's browser IDE

### DISCUSSION THREAD

**Discussion by SugarRayLua** [lyte2d/lyte2d#60] [Talk]
At least for mobile devices that would make it easy to program with lyte2D directly on the mobile device. The amulet browser IDE even lets one export the code as .wasm directly from the IDE in the browser:

https://www.amulet.xyz/editor.html

Fyi, and thanks for considering.

**Comment by morew4rd** [comment]
This is a cool idea, and something like this is on the _backlog_. 

I see two approaches here. I'm leaning towards (1) and Amulet and some others are like (2).

1. Write the editor in Lyte2D itself (think how Godot's editor is written in Godot using it's UI facilities.)  This can be made to work in WASM easily. And the editor/plugins would be in Lua.
2. Use an established open source web IDE (I'm thinking Code Mirror 6.) There are examples and should be doable.

Perhaps both of these can be done and may have their places. Also note that both of these may require some features from Lyte2D itself.

I'm thinking perhaps (2) can be approached by community as well. Perhaps someone can take an initial stab at this one. If you have interest/time in something like this, let me know!

**Comment by morew4rd** [comment]
A super cool bonus for (2) would be to integrate it with Lyte2D website itself (whose source code is also in this repo) so that live editing of samples would be possible.

**Comment by SugarRayLua** [comment]
Thanks for considering this! I would like to help but don't think I have the programming skill to do so yet. I'll keep working on my programming ability and keep a lookout for how/when I can contribute :-)

**Comment by lplume** [comment]
Hello! Slightly related, I wanted to try out to develop a simple IMGUI for lyte. Slowly but surely and I really do not know much about gui. Any tips and resources are welcome. I've got some starting code at this gist [starting idea](https://gist.github.com/lplume/afa9e0b933f06deb8b8d1d573a7933af)

**Comment by morew4rd** [comment]
@lplume very cool! I don't have much guidance, other than to say that I have used love2d library SUIT in the past and it was fine. But I don't have much other experience with imguis.

Please let me know if you hit issues, or find something that should exists but doesn't in the API.

**Comment by lplume** [comment]
nice to point out SUIT! At a first glance looks like I got more or less the same idea! Stay tuned! More to come

**Comment by morew4rd** [comment]
Based on the chat with @lplume on #40, I'm adding some ideas.

**Editor-as-a-lyte2d-app thoughts (from requirements perspective):**
- Filesystem access style: rxi's "lite" (current folder only) or "litexl" (everything)
  - For former, more PHYSFS functions can be exposed. (`ls`, `mkdir`, is_directory(), etc.)
  - For latter, we need either LFS or implement some of it ourselves
  - Do we need a "native" load/save dialog support?
- How does font rendering really look? Is lyte2d's font rendering good enough for long term text? Some testing and comparison may be needed. Comparison to love2d, lite and litexl would be interesing.
- Lyte2D currently doesn't expose events directly. "You check something if you care about it."  Is this a problem for UI tools? I don't think so, but I'm open to thinking more about this
- Is Lyte2D's input functionality good enough for this? See #39 for example.

Some for from Lyte2D will be needed for this in most cases. Let's identify these.

Thoughts?

**Comment by morew4rd** [comment]
Also slightly related to font issue: https://github.com/orgs/lyte2d/discussions/16

**Comment by lplume** [comment]
My bits:

- filesystem access: about the editor itself it could benefit some sort of filesystem access, i would suggest to think about "would lyte2d benefits from filesystem access?", at the moment i have no idea. It is possible to access file via lua's io, i do not think about directory listing; thinking of an editor it sure could use it.
- events and UI's: well... there are no ideal solution. If lyte2d do not fire events directly a library/wrapper (in pure lua probably) could be written to manage events. Im not so deep into lyte2d dev'ing,  might change my mind but so far i like how it is shaped.
Constraints usually opens to creativity, i'd like to hear from other dev who's actually experimenting the engine to gamedev

**Comment by morew4rd** [comment]
Converted to a github discussion. Any specific work item that may come out can be tracked as issues.
